src/dataflow/decode/mod.rs

```diff
@@ -118,6 +118,344 @@ where
+        result.push(Datum::from(bytes));
+        Ok(result.finish())
+    }
+
+    /// give a session a key-value pair
+    fn give_key_value<'a>(
+        &mut self,
+        key: Row,
+        bytes: &[u8],
+        line_no: &Option<i64>,
+        session: &mut PushSession<'a, (Row, Option<Row>)>,
+        time: Timestamp,
+    ) {
+        session.give((
+            (key, Some(pack_with_line_no(Datum::from(bytes), line_no))),
+            time,
+            1,
+        ));
+    }
+
+    /// give a session a plain value
+    fn give_value<'a>(
+        &mut self,
+        bytes: &[u8],
+        line_no: &Option<i64>,
+        session: &mut PushSession<'a, Row>,
+        time: Timestamp,
+    ) {
+        session.give((pack_with_line_no(Datum::from(bytes), line_no), time, 1));
+    }
+
+    /// Register number of success and failures with decoding
+    fn log_error_count(&self) {}
+}
+
+pub struct TextDecoderState;
+
+impl DecoderState for TextDecoderState {
+    fn reset_event_count(&mut self) {}
+
+    fn decode_key(&mut self, bytes: &[u8]) -> Result<Row, String> {
+        let mut result = RowPacker::new();
+        result.push(Datum::from(std::str::from_utf8(&bytes).ok()));
+        Ok(result.finish())
+    }
+
+    /// give a session a key-value pair
+    fn give_key_value<'a>(
+        &mut self,
+        key: Row,
+        bytes: &[u8],
+        line_no: &Option<i64>,
+        session: &mut PushSession<'a, (Row, Option<Row>)>,
+        time: Timestamp,
+    ) {
+        session.give((
+            (
+                key,
+                Some(pack_with_line_no(
+                    Datum::from(std::str::from_utf8(bytes).ok()),
+                    line_no,
+                )),
+            ),
+            time,
+            1,
+        ));
+    }
+
+    /// give a session a plain value
+    fn give_value<'a>(
+        &mut self,
+        bytes: &[u8],
+        line_no: &Option<i64>,
+        session: &mut PushSession<'a, Row>,
+        time: Timestamp,
+    ) {
+        session.give((
+            pack_with_line_no(Datum::from(std::str::from_utf8(bytes).ok()), line_no),
+            time,
+            1,
+        ));
+    }
+
+    /// Register number of success and failures with decoding
+    fn log_error_count(&self) {}
+}
+
+fn decode_kafka_upsert_inner<G, K: 'static, V: 'static>(
+    stream: &Stream<G, ((Vec<u8>, Vec<u8>), Option<i64>)>,
+    mut key_decoder_state: K,
+    mut value_decoder_state: V,
+    op_name: &str,
+) -> Stream<G, ((Row, Option<Row>), Timestamp, Diff)>
+where
+    G: Scope<Timestamp = Timestamp>,
+    K: DecoderState,
+    V: DecoderState,
+{
+    stream.unary(
+        Exchange::new(|x: &((Vec<u8>, _), _)| (x.0).0.hashed()),
+        &op_name,
+        move |_, _| {
+            move |input, output| {
+                input.for_each(|cap, data| {
+                    let mut session = output.session(&cap);
+                    for ((key, payload), aux_num) in data.iter() {
```

Comment:
this might be a dumb question but is `aux_num` meant to be the global index of this record (relative to the total count of records ever received) or the local index of this record in this batch? It's unclear what it is based on this code, but my understanding given pack_with_line_no is that its meant to be the former?

Reply:
Based on reading the code, it is the global index of this record, though the `aux_num` is None for Kinesis. 

I guess I put it as `aux_num` because I myself was not too sure what this number does and I figure I should allow for flexibility in what it stands for.